feat: add command custom fields (#1158)
canvas-medical/documentation: collections/_sdk/effects/command_metadata_create_form.md

---
title: "Command Metadata Create form"
slug: "command-metadata-create-form-effect"
excerpt: "Effect for dynamically displaying additional fields on a command"
hidden: false
---

## Overview

> This allows developers to dynamically display additional fields on a command. The values entered in these fields are stored as [command metadata](/sdk/data-command/#commandmetadata) against the target `command_uuid`.

Reviewer: ```suggestion
The `CommandMetadataCreateFormEffect` allows developers to dynamically display additional fields with a command in a note. The values entered in these fields are stored as [command metadata](/sdk/data-command/#commandmetadata) against the target `command_uuid`.
```

| `key`              | `str`                | unique identifier of the field - command metadata key             |
| `label`            | `str`                | the label that will be displayed on the field                     |
| `type`             | `InputType`          | the type of the input - TEXT, SELECT, DATE.                       |
| `required`         | `bool`               | if the input is required.                                         |
| `editable`         | `bool`               | if the input can be editabled.                                    |
| `options`          | `list[str]`          | possible options for when the input type is set to "SELECT”       |
| `value`            | `str`                | default value for the field                                       |

### **CommandMetadataCreateFormEffect**

A CommandMetadataCreateFormEffect consists of the following properties:

#### Attributes

| Attribute          | Type              | Description                                                       |
|--------------------|-------------------|-------------------------------------------------------------------|

> | `command_uuid`     | `str`             | the UUID of the command these fields should be rendered on.       |

Reviewer: I could see a world where for the `CommandMetadataCreateFormEffect` you just always want specific fields per schema key. 

I was thinking that this would be similar to the Patient metadata. When you are setting the metadata, you are supplying the patient you are updating. But the `PatientMetadataCreateFormEffect` effect doesnt require the patient id because its a form that is global. 

I would think when I am rendering the command fields, its not specific to that command uuid, its more specific to that schema_key. But it does look like the schema_key is in the self.context object so maybe we just want the examples on this page to check the schema_key

Author: I’ll update the examples based on your suggestion 👍  

We did it this way because it could be useful to let the developer decide whether to apply the logic to all schema keys or only to specific ones on certain notes, for example.

| `form_fields`      | `list[FormField]` | list of fields.                                                   |

## Validation

The effect validates inputs before it is applied:

- `command_uuid` is required.
- `options` may only be set on fields whose `type` is `InputType.SELECT`; providing `options` on a `TEXT` or `DATE` field raises a validation error.
- Every `key` must be unique across `form_fields`. Duplicates raise a validation error per duplicated key.

## Example Usage

The following handler declares two extra fields on every plan command when the platform requests additional fields for it:

```python
from canvas_sdk.effects import Effect
from canvas_sdk.effects.command_metadata import (
    CommandMetadataCreateFormEffect,
    FormField,
    InputType,
)
from canvas_sdk.events import EventType
from canvas_sdk.handlers import BaseHandler

class PlanCommandAdditionalFields(BaseHandler):
    RESPONDS_TO = EventType.Name(EventType.COMMAND__FORM__GET_ADDITIONAL_FIELDS)

    def compute(self) -> list[Effect]:
        form = CommandMetadataCreateFormEffect(

>             command_uuid=self.event.target.id,

Reviewer: I thought the self.context has the command_uuid in it. There is no documentation of what the self.event.target is

<img width="909" height="181" alt="Image" src="https://github.com/user-attachments/assets/c2e5d21a-0c20-4307-8be1-d6757f4bbbad" />

Author: You’re correct, the command_uuid is in the target, so we don’t need to duplicate it in the context

            form_fields=[
                FormField(
                    key="priority",
                    label="Priority",
                    type=InputType.SELECT,
                    options=["low", "medium", "high"],
                ),
                FormField(
                    key="follow_up_date",
                    label="Follow-up date",
                    type=InputType.DATE,
                    editable=True,
                ),
            ],
        )

        return [form.apply()]
```

> Once the user fills out these fields, their values are persisted as command metadata and can be read back using the standard [command metadata](/sdk/data-command/#commandmetadata) APIs.

Reviewer: ```suggestion
Once the user fills out these fields, their values are persisted as command metadata and can be read back through the SDK [command metadata](/sdk/data-command/#commandmetadata) table.
```
